### PR TITLE
[video] Separate video versions and extras items

### PR DESCRIPTION
**Pull request by ksooo**
Fixes the last open problem reported in issue #24245 : "the extras rename selection dialog should contain items specific to extras, same as the wording of the window heading and its buttons". 

All lists (not only on rename, like requested in the issue) are now only displaying the items belonging to the respective asset type.

For this we need to extend the database scheme, because in the original design the entries in the "videoversiontype" table were not for specific assets (versions, extras, ...), but for all types of assets.

I tried to migrate data as good as possible, by inspecting where the types are actually used, to find out the asset type for those types. That was actually the most challenging (== time consuming) part of this PR.

Runtime-tested on macOS, latest Kodi master.

@enen92 can you please have a look?

### CODE REVIEW

File: xbmc/video/VideoDatabase.cpp
Comment (enen92) [inline xbmc/video/VideoDatabase.cpp:11749]: Unsigned int?

**Review by enen92 — APPROVED** [review]
Looks good as far as I can tell. Probably better ask someone for the database changes, not something I play with often.

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11749]: > Unsigned int?

Logically yes, but wanted to be consistent at least => `int CVideoDatabase::GetSchemaVersion() const`

Comment (enen92) [reply xbmc/video/VideoDatabase.cpp:11749]: Right

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:6211]: I understand what you're going for, but I don't think it's possible for the situation to happen?
The code prior to this PR (up to db v126) precreated all records from id 40400 to 40800, reserving them for versions (therefore excluding extras from that range).
Unless I missed a place in code where extras management code reuses an id between 40400 to 40800 instead of allocating a new id.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:6211]: Rename an extra, pick a system provided version, like „standard edition“ and there you go. This is very well possible, unfortunately. I tried that out.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:6211]: > precreated all records from id 40400 to 40800, reserving them for versions (therefore excluding extras from that range).

The last part of your statement is wrong. Before this PR, the types were agnostic to versions / extras. That’s the actual problem the PR fixed. The reservation was not for versions, only for „system provided“.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:11761]: A better change would be to just replace the SQL with this:
`INSERT INTO videoversiontype (id, name, owner) VALUES (%i, '%s', %i)`
It's bad practice to not specify the column names. and causes issues when the table definition changes.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11761]: Oh yes. I will change that.

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11761]: Hold on, you were right, a change is needed for the new install case, where no migration is going to take place.
So either what you did, or what I suggested + a DEFAULT value for the column.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11761]: Sorry, I do not understand. Post a diff please.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11761]: ""INSERT INTO videoversiontype (id, name, owner, itemType) VALUES(NULL, '%s', %i, %i)" seems to work just fine. I actually copied that from existing code.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:11815]: Although unlikely in real life, if the user types a version name for the extra (Standard Edition), then this will find the existing "system" record for the version and update it/use it, instead of looking exclusively at extras?
So maybe `WHERE name = 'new name' AND itemtype = assetType`?
Maybe the 'owner' field should also be factored in this, but I don't have a good mental picture of its usage.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11761]: Yes it works but that wasn't the issue.
For a brand new installation, the migration code assigning itemType = VideoAssetType::VERSION will not be executed so this function needs to provide the value during the INSERT. otherwise the field will be NULL.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11761]: Not sure we have an issue here with ne install. I think I covered this case. Looking at CreateTables:

```c++
  m_pDS->exec("CREATE TABLE videoversiontype (id INTEGER PRIMARY KEY, name TEXT, owner INTEGER, "
              "itemType INTEGER)");
  InitializeVideoVersionTypeTable(GetSchemaVersion());
```

Important and new with this PR is passing the schema version number to `InitializeVideoVersionTypeTable`. Looking at changed implementation for `InitializeVideoVersionTypeTable`:

```c++
    for (int id = VIDEO_VERSION_ID_BEGIN; id <= VIDEO_VERSION_ID_END; ++id)
    {
      const std::string type{g_localizeStrings.Get(id)};
      if (schemaVersion < 127)
      {
        m_pDS->exec(
            PrepareSQL("INSERT INTO videoversiontype (id, name, owner) VALUES(%i, '%s', %i)", id,
                       type.c_str(), VideoAssetTypeOwner::SYSTEM));
      }
      else
      {
        m_pDS->exec(PrepareSQL(
            "INSERT INTO videoversiontype (id, name, owner, itemType) VALUES(%i, '%s', %i, %i)", id,
            type.c_str(), VideoAssetTypeOwner::SYSTEM, VideoAssetType::VERSION));
      }
    }
```

You see that itemType will be initialised properly, if the column is present (version > 127).

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11761]: yes that's fine.

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11815]: Right. Thanks. I will change that.

Comment (ksooo) [reply xbmc/video/VideoDatabase.cpp:11815]: > Maybe the 'owner' field should also be factored in this, but I don't have a good mental picture of its usage.

What I found out so far:

owner = SYSTEM 
=> predefined

owner = AUTO 
=> created without user interaction (during scanning? not sure),
=> On "add extra" (maybe also on "add version", you will not asked for selecting a type, the new extra will be created with an auto-created name. You have to use "Rename" later if you want to assign an existing type. Weird.

owner = USER
=> created by the user

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:11815]: OK it shouldn't be necessary to include owner in the WHERE at this time, existing logic seems fine.

Auto is the value I wasn't too sure of. Yes it comes from scraper and the add buttons. Auto names are not displayed in the rename lists, but renaming something to an existing auto name will "upgrade" it to "user" owner, visible in future rename lists. The idea is sound but I didn't inspect the code for deviations.

**Comment by ksooo** [comment]
Rebased and incorporated requested changes. Will merge once Jenkins is green.

**Comment by ksooo** [comment]
Jenkins build this please

**Comment by ksooo** [comment]
Please don't merge (yet). I want to rethink this...

**Comment by Ch1llb0** [comment]
As this was adressing one of the issue I reported in [#24245](https://github.com/xbmc/xbmc/issues/24245), I'd like to add some feedback after implementing the changes here: The assets are indeed now seperated and things look and feel a lot more perfected.

But one issue that's remaining from my issue report is this:
> same as the wording of the window heading and its buttons

The dialog heading as well as buttons when choosing to rename or add an extra are still referring to "versions":

- when choosing to add an extra:

screenshot00000

- when choosing to rename an extra:

screenshot00001

**Comment by ksooo** [comment]
@Ch1llb0 thanks for the feedback. Very valid point. May I suggest that you stop hijacking closed PRs and start writing issues?

**Comment by ksooo** [comment]
Please open a new issue, thanks.

**Comment by Ch1llb0** [comment]
Will do once I've got a moment 👍🏼

**Pull request merged**
